File: backend/app/services/public_host.py
"""Normalize the Settings public domain into a FarmOS host (farm.example.com)."""

from __future__ import annotations

import ipaddress
import re
from dataclasses import dataclass
from urllib.parse import urlparse

LABEL_RE = re.compile(r"^(?:xn--)?[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?$", re.I)
LOCAL_HOSTS = {"localhost", "localhost."}


@dataclass(frozen=True)
class PublicHost:
    raw: str
    host: str = ""
    origin: str = ""
    error: str = ""

    @property
    def ok(self) -> bool:
        return not self.error


EMPTY = PublicHost(raw="", host="", origin="")

_JUNK = "That doesn’t look like a domain or IP. Try example.com, farm.example.com, or a LAN address."


def normalize_public_host(raw: str | None) -> PublicHost:
    """Turn operator input into a public FarmOS origin, or an empty result if unset.

    Blank → no public host (local QR payloads stay farmos:kind:token).
    example.com → https://farm.example.com
    farm.example.com / https://farm.example.com/path → do not double-prefix.
    www.example.com → farm.example.com unless farm. was already present.
    localhost and IPs stay as typed (no farm. prefix).
    """
    text = (raw or "").strip()
    if not text:
        return EMPTY
    if any(ch.isspace() for ch in text):
        return PublicHost(raw=text, host="", origin="", error="Domain cannot contain spaces. Use example.com or farm.example.com.")
    if any(ch in text for ch in ("\\", "<", ">", '"', "'", "`", "@")):
        return PublicHost(raw=text, host="", origin="", error=_JUNK)
    try:
        scheme_in, host_in, port_in = _parse_input(text)
    except ValueError as exc:
        return PublicHost(raw=text, host="", origin="", error=str(exc) or _JUNK)

    host_in = (host_in or "").rstrip(".").lower()
    if not host_in:
        return PublicHost(raw=text, host="", origin="", error=_JUNK)
    if len(host_in) > 253:
        return PublicHost(raw=text, host="", origin="", error="That hostname is too long.")

    local = _is_local_or_ip(host_in)
    if not local:
        try:
            host_in = host_in.encode("idna").decode("ascii")
        except UnicodeError:
            return PublicHost(raw=text, host="", origin="", error=_JUNK)
        err = _validate_hostname(host_in)
        if err:
            return PublicHost(raw=text, host="", origin="", error=err)

    host = host_in if local else _apply_farm_prefix(host_in)
    if not local:
        err = _validate_hostname(host)
        if err:
            return PublicHost(raw=text, host="", origin="", error=err)

    if not local:
        scheme = "https"
    elif scheme_in == "https":
        scheme = "https"
    else:
        scheme = "http"

    origin = f"{scheme}://{_hostport(host, port_in, scheme)}"
    return PublicHost(raw=text, host=_hostport(host, port_in, scheme), origin=origin)


def public_origin_or_empty(raw: str | None) -> str:
    result = normalize_public_host(raw)
    if not result.ok:
        raise ValueError(result.error)
    return result.origin


def _parse_input(text: str) -> tuple[str, str, int | None]:
    if "://" in text:
        parsed = urlparse(text)
        scheme = (parsed.scheme or "").lower()
        if scheme not in {"http", "https"}:
            raise ValueError("Use http:// or https://, or enter a domain with no scheme.")
    else:
        body = text.lstrip("/")
        if _bare_ipv6(body):
            body = f"[{body}]"
        parsed = urlparse(f"//{body}")
        scheme = ""
    if parsed.username or parsed.password:
        raise ValueError(_JUNK)
    try:
        port = parsed.port
    except ValueError as exc:
        raise ValueError("That port number is not valid.") from exc
    host = parsed.hostname
    if not host:
        raise ValueError(_JUNK)
    return scheme, host, port


def _bare_ipv6(text: str) -> bool:
    head = text.split("/")[0]
    if head.startswith("["):
        return False
    candidate = head.split("%")[0]
    try:
        ipaddress.IPv6Address(candidate)
        return True
    except ValueError:
        return False


def _is_local_or_ip(host: str) -> bool:
    if host in LOCAL_HOSTS or host.endswith(".localhost"):
        return True
    try:
        ipaddress.ip_address(host)
        return True
    except ValueError:
        return False


def _apply_farm_prefix(host: str) -> str:
    labels = [p for p in host.split(".") if p]
    if not labels:
        return host
    if labels[0] == "www" and len(labels) > 1:
        labels = labels[1:]
        host = ".".join(labels)
    if labels[0] == "farm":
        return host
    if "." not in host:
        # Single-label LAN names stay as-is — do not invent farm.localhost-style hosts.
        return host
    return "farm." + host


def _validate_hostname(host: str) -> str:
    labels = host.split(".")
    if any(not label for label in labels):
        return "That domain has an empty label (check the dots)."
    if any(not LABEL_RE.match(label) for label in labels):
        return _JUNK
    return ""


def _hostport(host: str, port: int | None, scheme: str) -> str:
    try:
        ip = ipaddress.ip_address(host)
        display = f"[{host}]" if ip.version == 6 else host
    except ValueError:
        display = host
    default = 443 if scheme == "https" else 80
    if port and port != default:
        return f"{display}:{port}"
    return display

File: frontend/lib/public-host.ts
/** Normalize Settings → Public domain. Keep in sync with backend/app/services/public_host.py. */

export type PublicHost = {
  raw: string;
  host: string;
  origin: string;
  error: string;
};

const JUNK =
  "That doesn’t look like a domain or IP. Try example.com, farm.example.com, or a LAN address.";

const LABEL = /^(?:xn--)?[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?$/i;

export function normalizePublicHost(raw: string | null | undefined): PublicHost {
  const text = (raw || "").trim();
  if (!text) return { raw: "", host: "", origin: "", error: "" };
  if (/\s/.test(text)) {
    return {
      raw: text,
      host: "",
      origin: "",
      error: "Domain cannot contain spaces. Use example.com or farm.example.com.",
    };
  }
  if (/[\\<>"'`@]/.test(text)) {
    return { raw: text, host: "", origin: "", error: JUNK };
  }

  let schemeIn = "";
  let hostname = "";
  let port: number | null = null;
  try {
    const parsed = parseInput(text);
    schemeIn = parsed.scheme;
    hostname = parsed.hostname;
    port = parsed.port;
  } catch (err) {
    return { raw: text, host: "", origin: "", error: err instanceof Error ? err.message : JUNK };
  }

  hostname = hostname.replace(/\.+$/, "").toLowerCase();
  if (!hostname) return { raw: text, host: "", origin: "", error: JUNK };
  if (hostname.length > 253) {
    return { raw: text, host: "", origin: "", error: "That hostname is too long." };
  }

  const local = isLocalOrIp(hostname);
  if (!local) {
    const err = validateHostname(hostname);
    if (err) return { raw: text, host: "", origin: "", error: err };
  }

  const host = local ? hostname : applyFarmPrefix(hostname);
  if (!local) {
    const err = validateHostname(host);
    if (err) return { raw: text, host: "", origin: "", error: err };
  }

  let scheme: string;
  if (schemeIn === "http" || schemeIn === "https") {
    scheme = local ? schemeIn : "https";
  } else {
    scheme = local ? "http" : "https";
  }

  const hostport = formatHostPort(host, port, scheme);
  return { raw: text, host: hostport, origin: `${scheme}://${hostport}`, error: "" };
}

function parseInput(text: string): { scheme: string; hostname: string; port: number | null } {
  let candidate: string;
  let scheme = "";
  if (text.includes("://")) {
    const schemePart = text.split("://", 1)[0]?.toLowerCase() || "";
    if (schemePart !== "http" && schemePart !== "https") {
      throw new Error("Use http:// or https://, or enter a domain with no scheme.");
    }
    scheme = schemePart;
    candidate = text;
  } else {
    let body = text.replace(/^\/+/, "");
    if (isBareIPv6(body)) body = `[${body}]`;
    candidate = `http://${body}`;
  }
  let url: URL;
  try {
    url = new URL(candidate);
  } catch {
    throw new Error(JUNK);
  }
  if (url.username || url.password) throw new Error(JUNK);
  const hostname = url.hostname.replace(/^\[/, "").replace(/]$/, "");
  if (!hostname) throw new Error(JUNK);
  let port: number | null = null;
  if (url.port) {
    const n = Number(url.port);
    if (!Number.isInteger(n) || n < 1 || n > 65535) {
      throw new Error("That port number is not valid.");
    }
    port = n;
  }
  return { scheme, hostname, port };
}

function isBareIPv6(text: string): boolean {
  const head = text.split("/")[0] || "";
  if (head.startsWith("[")) return false;
  return isIPv6(head.split("%")[0] || "");
}

function isLocalOrIp(host: string): boolean {
  if (host === "localhost" || host.endsWith(".localhost")) return true;
  return isIPv4(host) || isIPv6(host);
}

function applyFarmPrefix(host: string): string {
  let labels = host.split(".").filter(Boolean);
  if (!labels.length) return host;
  if (labels[0] === "www" && labels.length > 1) {
    labels = labels.slice(1);
    host = labels.join(".");
  }
  if (labels[0] === "farm") return host;
  if (!host.includes(".")) return host;
  return `farm.${host}`;
}

function validateHostname(host: string): string {
  const labels = host.split(".");
  if (labels.some((label) => !label)) return "That domain has an empty label (check the dots).";
  if (labels.some((label) => !LABEL.test(label))) return JUNK;
  return "";
}

function formatHostPort(host: string, port: number | null, scheme: string): string {
  const display = isIPv6(host) ? `[${host}]` : host;
  const defaultPort = scheme === "https" ? 443 : 80;
  if (port && port !== defaultPort) return `${display}:${port}`;
  return display;
}

function isIPv4(host: string): boolean {
  const parts = host.split(".");
  if (parts.length !== 4) return false;
  return parts.every((p) => {
    if (!/^\d{1,3}$/.test(p)) return false;
    const n = Number(p);
    return n >= 0 && n <= 255 && String(n) === p;
  });
}

function isIPv6(host: string): boolean {
  if (!host.includes(":")) return false;
  // URL constructor already accepted it; a colon-separated host that is not a hostname label.
  if (/^[0-9a-f:.]+$/i.test(host) && /:/.test(host)) return true;
  return false;
}
